hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenMintHandler.java

```diff
@@ -54,8 +77,174 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
+     * @param treasuryRel   - the treasury relation of the token
+     * @param metadata      - the metadata of the nft to be minted
+     * @param consensusTime - the consensus time of the transaction
+     * @param accountStore  - the account store
+     * @param tokenStore    - the token store
+     * @param tokenRelStore - the token relation store
+     * @param nftStore      - the nft store
+     */
+    private List<Long> mintNonFungible(
+            final Token token,
+            @NonNull final TokenRelation treasuryRel,
+            @NonNull final List<Bytes> metadata,
+            @NonNull final Instant consensusTime,
+            @NonNull final WritableAccountStore accountStore,
+            @NonNull final WritableTokenStore tokenStore,
+            @NonNull final WritableTokenRelationStore tokenRelStore,
+            @NonNull final WritableNftStore nftStore) {
+        final var metadataCount = metadata.size();
+        validateFalse(metadata.isEmpty(), INVALID_TOKEN_MINT_METADATA);
+
+        // validate token number from treasury relation
+        final var tokenId = asToken(treasuryRel.tokenNumber());
+        validateTrue(treasuryRel.tokenNumber() == token.tokenNumber(), FAIL_INVALID);
+
+        // get the treasury account
+        final var treasuryAccount = accountStore.get(asAccount(treasuryRel.accountNumber()));
+        validateTrue(treasuryAccount != null, INVALID_TREASURY_ACCOUNT_FOR_TOKEN);
+
+        // get the latest serial number minted for the token
+        var currentSerialNumber = token.lastUsedSerialNumber();
+        validateTrue((currentSerialNumber + metadataCount) <= MAX_NUM_ALLOWED, SERIAL_NUMBER_LIMIT_REACHED);
+
+        // Change the supply on token
+        changeSupply(token, treasuryRel, metadataCount, FAIL_INVALID, accountStore, tokenStore, tokenRelStore);
+
+        final var mintedSerials = new ArrayList<Long>(metadata.size());
+
+        // for each serial number minted increment serial numbers and create new unique token
+        for (final var meta : metadata) {
+            currentSerialNumber++;
+            // The default sentinel account is used (0.0.0) to represent unique tokens owned by the treasury
+            final var uniqueToken =
+                    buildNewlyMintedNft(treasuryAccount, consensusTime, tokenId, meta, currentSerialNumber);
+            nftStore.put(uniqueToken);
+            // all minted serials should be added to the receipt
+            mintedSerials.add(currentSerialNumber);
+        }
+        // Update last used serial number and number of owned nfts and put the updated token and treasury
+        // into the store
+        final var copyToken = token.copyBuilder();
+        final var copyTreasury = treasuryAccount.copyBuilder();
+
+        copyToken.lastUsedSerialNumber(currentSerialNumber);
+        copyTreasury.numberOwnedNfts(treasuryAccount.numberOwnedNfts() + metadataCount);
+        copyTreasury.headNftSerialNumber(currentSerialNumber);
```

Comment:
We actually _don't_ track treasury-owned NFTs in these linked lists, because their purpose is only to return to treasury the NFTs owned by a expired account. (So an NFT already owned by the treasury doesn't need this support.)

The `mono-service` code is a bit mysterious I admit, but you can see [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ledger/interceptors/LinkAwareUniqueTokensCommitInterceptor.java#L98) that links are only updated for a new NFT if its initial owner **isn't** the `0.0.0` treasury sentinel.

(And note we if a non-fungible token type expires, we can just iterate through its serial numbers `0, 1, 2...` and clean it up that way. So there's no need to "link" treasury-owned NFTs in either case.)

Reply:
Thanks Michael. 🙌  Will fix this in my next PR